controller/setting_controller.go

```diff
@@ -1023,7 +1126,70 @@ func (sc *SettingController) updateNodeSelector() error {
 				}
 			}
 		default:
-			return fmt.Errorf("unknown object type %v when updating %v setting", objType, types.SettingNamePriorityClass)
+			return fmt.Errorf("unknown object type %v when updating %v setting", objType, types.SettingNameSystemManagedComponentsNodeSelector)
+		}
+	}
+
+	return nil
+}
+
+// updateCSISidecarComponentsNodeSelector deletes all user-deployed and system-managed components immediately with the updated node selector.
+func (sc *SettingController) updateCSISidecarComponentsNodeSelector() error {
+	setting, err := sc.ds.GetSettingWithAutoFillingRO(types.SettingNameSystemManagedCSISidecarComponentsNodeSelector)
+	if err != nil {
+		return err
+	}
+	newNodeSelector, err := types.UnmarshalNodeSelector(setting.Value)
+	if err != nil {
+		return err
+	}
+
+	updatingRuntimeObjects, err := sc.collectRuntimeObjectsKubernetesCSI()
+	if err != nil {
+		return errors.Wrap(err, "failed to collect runtime objects for node selector update")
+	}
+	notUpdatedNodeSelectorObjs, err := getNotUpdatedNodeSelectorList(newNodeSelector, updatingRuntimeObjects...)
+	if err != nil {
+		return err
+	}
+	if len(notUpdatedNodeSelectorObjs) == 0 {
+		return nil
+	}
+
+	detached, err := sc.ds.AreAllVolumesDetachedState()
+	if err != nil {
+		return errors.Wrapf(err, "failed to check volume detachment for %v setting update", types.SettingNameSystemManagedCSISidecarComponentsNodeSelector)
+	}
+	if !detached {
+		return &types.ErrorInvalidState{Reason: fmt.Sprintf("failed to apply %v setting to Longhorn components when there are attached volumes. It will be eventually applied", types.SettingNameSystemManagedCSISidecarComponentsNodeSelector)}
+	}
+
+	for _, obj := range notUpdatedNodeSelectorObjs {
+		switch objType := obj.(type) {
+		case *appsv1.DaemonSet:
+			ds := obj.(*appsv1.DaemonSet)
+			sc.logger.Infof("Updating the node selector from %v to %v for %v", ds.Spec.Template.Spec.NodeSelector, newNodeSelector, ds.Name)
+			ds.Spec.Template.Spec.NodeSelector = newNodeSelector
+			if _, err := sc.ds.UpdateDaemonSet(ds); err != nil {
```

Comment:
I think casting isn’t needed. The type switch already handles that. [https://go.dev/ref/spec#Type_switches](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)

```suggestion
        switch objTyped := obj.(type) {
        case *appsv1.DaemonSet:
            sc.logger.Infof("Updating the node selector from %v to %v for %v", objTyped.Spec.Template.Spec.NodeSelector, newNodeSelector, ds.Name)
            objTyped.Spec.Template.Spec.NodeSelector = newNodeSelector
            if _, err := sc.ds.UpdateDaemonSet(objTyped); err != nil {
```